core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -885,13 +885,25 @@ public CompletionStats completionStats(String... fields) {
     }
 
     /**
-     * Rolls the tranlog generation.
+     * checks and removes translog files that no longer need to be retained. See
+     * {@link org.elasticsearch.index.translog.TranslogDeletionPolicy} for details
+     */
+    public void trimTranslog() {
+        verifyNotClosed();
+        final Engine engine = getEngine();
+        engine.trimTranslog();
+    }
+
+    /**
+     * Rolls the tranlog generation and cleans unneeded.
      *
      * @throws IOException if any file operations on the translog throw an I/O exception
      */
-    private void rollTranslogGeneration() throws IOException {
+    private void rollTranslogAndTrimGeneration() throws IOException {
         final Engine engine = getEngine();
-        engine.getTranslog().rollGeneration();
+        final Translog translog = engine.getTranslog();
+        translog.rollGeneration();
+        translog.trimUnreferencedReaders();
```

Comment:
I think the readlock is healthy here, I wonder if we would rather fold this part into the engine and don't call it from the outside and add a parameter `boolean rollGeneration` to the trim method?